File: cpp/v001_original.cpp
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

typedef PRECISION real;

const int NX = 128;
const int NY = 128;
const int MAX_ITERATIONS = 1<<16;

inline int idx(int i, int j) {return j + i*NY;}

real* make_array(int nx, int ny) {
  return new real[nx*ny];
}

void run_jacobi(real *p, const real* b, const real dx, const real dy) {
  real D = 2.0*(dx*dx + dy*dy);
  real D_x = dy*dy/D;
  real D_y = dx*dx/D;
  real B = -(dx*dx*dy*dy)/D;

  real* p_new = make_array(NX, NY);
  for(int iter = 0; iter<MAX_ITERATIONS; ++iter) {
    for(int i=1; i<NX-1; ++i) {
      for(int j=1; j<NX-1; ++j) {
        p_new[idx(i,j)] = D_x*(p[idx(i+1,j)] + p[idx(i-1,j)]) + D_y*(p[idx(i,j+1)] + p[idx(i,j-1)]) + B*b[idx(i,j)];
      }
    }
    real* temp = p_new;
    p_new = p;
    p = temp;
  }
  delete[] p_new;
}

int main(int argc, char* argv[]) {
  real* p = make_array(NX, NY);
  real* b = make_array(NX, NY);
  real* p_soln = make_array(NX, NY);

  real dx = 1.0/(NX-1);
  real dy = 1.0/(NY-1);

  for(int i=0; i<NX; ++i) {
    for(int j=0; j<NY; ++j) {
      real x = i*dx;
      real y = j*dx;

      b[idx(i,j)] = sin(M_PI*x)*sin(M_PI*y);
      p_soln[idx(i,j)] = -sin(M_PI*x)*sin(M_PI*y)/(2.0*M_PI*M_PI);
      p[idx(i,j)] = 0.0;
    }
  }

  clock_t start = clock();
  run_jacobi(p, b, dx, dy);
  clock_t diff = clock() - start;

  int msec = diff * 1000 / CLOCKS_PER_SEC;

  real av_error = 0.0;
  for(int i=1; i<NX-1; ++i) {
    for(int j=1; j<NY-1; ++j) {
      av_error += fabs(p[idx(i,j)] - p_soln[idx(i,j)]);
    }
  }
  av_error /= (NX*NY);

  printf("%s, cpp, %d, %d, %d, %d, %e\n", argv[0], NX, NY, MAX_ITERATIONS, msec, av_error);

  delete[] p;
  delete[] b;
  delete[] p_soln;

  (void)argc; // Disable compiler warnings for argc
  return 0;
}

File: python/v001_original.py
import numpy as np
import os
import time

real = np.float64

NX = 128
NY = 128
MAX_ITERATIONS = 65536

DX = 1.0/(NX-1);
DY = 1.0/(NY-1);

D = 2.0*(DX*DX + DY*DY);
D_x = DY*DY/D;
D_y = DX*DX/D;
B = -(DX*DX*DY*DY)/D;

def main():
    p = np.zeros((NX, NY), dtype=real)
    b = np.zeros((NX, NY), dtype=real)
    p_soln = np.zeros((NX, NY), dtype=real)

    x = np.linspace(0.0, 1.0, num=NX)
    y = np.linspace(0.0, 1.0, num=NY)
    X, Y = np.meshgrid(x, y, indexing='ij')
    b[:,:] = np.sin(np.pi*X)*np.sin(np.pi*Y)
    p_soln[:,:] = -np.sin(np.pi*X)*np.sin(np.pi*Y)/(2.0*np.pi**2)

    start = time.perf_counter()
    p_new = np.zeros_like(p)
    for i in range(MAX_ITERATIONS):
        p_new[1:-1,1:-1] = D_x*(p[2:,1:-1] + p[:-2,1:-1]) + D_y*(p[1:-1,2:] + p[1:-1,:-2]) + B*b[1:-1,1:-1]
        p_new, p = p, p_new
    runtime = (time.perf_counter() - start)*1000

    av_error = np.sum(np.abs(p - p_soln))/(NX*NY)

    print(os.path.basename(__file__), "Python", NX, NY, MAX_ITERATIONS, runtime, av_error)


main()
